tls/s2n_key_log.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <s2n.h>
+#include "tls/s2n_config.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_crypto_constants.h"
+#include "tls/s2n_quic_support.h" /* this currently holds the s2n_secret_type_t enum */
+#include "utils/s2n_blob.h"
+#include "utils/s2n_safety.h"
+
+/* hex requires 2 chars per byte */
+#define HEX_ENCODING_SIZE 2
+
+/* https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Key_Log_Format */
+
+S2N_RESULT s2n_key_log_hex_encode(struct s2n_stuffer *output, uint8_t *bytes, size_t len)
+{
+    ENSURE_MUT(output);
+    ENSURE_REF(bytes);
+
+    const uint8_t chars[] = "0123456789abcdef";
+
+    for (size_t i = 0; i < len; i++) {
+        uint8_t upper = bytes[i] >> 4;
+        uint8_t lower = bytes[i] & 0x0f;
+
+        GUARD_AS_RESULT(s2n_stuffer_write_uint8(output, chars[upper]));
+        GUARD_AS_RESULT(s2n_stuffer_write_uint8(output, chars[lower]));
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_key_log_tls13_secret(struct s2n_connection *conn, struct s2n_blob *secret, s2n_secret_type_t secret_type)
+{
+    ENSURE_REF(conn);
+    ENSURE_REF(conn->config);
+    ENSURE_REF(secret);
+
+    /* only emit keys if the callback has been set */
+    if (!conn->config->key_log_cb) {
+        return S2N_RESULT_OK;
+    }
+
+    const uint8_t client_early_traffic_label[] = "CLIENT_EARLY_TRAFFIC_SECRET ";
+    const uint8_t client_handshake_label[] = "CLIENT_HANDSHAKE_TRAFFIC_SECRET ";
+    const uint8_t server_handshake_label[] = "SERVER_HANDSHAKE_TRAFFIC_SECRET ";
+    const uint8_t client_traffic_label[] = "CLIENT_TRAFFIC_SECRET_0 ";
+    const uint8_t server_traffic_label[] = "SERVER_TRAFFIC_SECRET_0 ";
+
+    const uint8_t *label = NULL;
+    uint8_t label_size = 0;
+
+    switch (secret_type) {
+        case S2N_CLIENT_EARLY_TRAFFIC_SECRET:
+            label = client_early_traffic_label;
```

Comment:
Would it be cleaner to have a lookup table of s2n_secret_type_t -> label, and then just call sizeof(label)?

Reply:
A lookup table wouldn't be able to handle enum values that we don't understand here. I think the switch works well.